File: project_euler/Problem031/Problem031.cpp
#include <iostream>
#include <vector>
using namespace std;

class Problem031 {
public:
    void solution() {
        int target = 200;
        vector<int> coins = {1, 2, 5, 10, 20, 50, 100, 200};
        vector<int> ways(target + 1, 0);
        ways[0] = 1;
        for (int coin : coins) {
            for (int i = coin; i <= target; ++i) {
                ways[i] += ways[i - coin];
            }
        }
        cout << ways[target] << endl;
    }
};

File: project_euler/Problem031/Problem031.java
package project_euler;

public class Problem031 {
    public void solution() {
        int target = 200;
        int[] coins = new int[]{1, 2, 5, 10, 20, 50, 100, 200};
        int[] ways = new int[target + 1];
        ways[0] = 1;
        for (int coin : coins) {
            for (int i = coin; i <= target; ++i) {
                ways[i] += ways[i - coin];
            }
        }
        System.out.println(ways[target]);
    }
}
